google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionImplTest.java

```diff
@@ -274,6 +276,49 @@ private static long utcTimeSeconds(int year, int month, int day, int hour, int m
     return calendar.getTimeInMillis() / 1000;
   }
 
+  @Test
+  public void multiplesSingleUseContext_whenMultiplexedSession_assertValidContext() {
+    final String dbName = "projects/p1/instances/i1/databases/d1";
+    final DatabaseId db = DatabaseId.of(dbName);
+    final String sessionName = dbName + "/sessions/s1";
+    final SpannerImpl spanner = new SpannerImpl(rpc, spannerOptions);
+    final SessionConsumer consumer =
+        new SessionConsumer() {
+          @Override
+          public void onSessionReady(SessionImpl session) {
+            ReadContext ctx1 = session.singleUse(TimestampBound.strong());
+            session.singleUse(TimestampBound.strong());
+            try {
+              ResultSet resultSet =
+                  ctx1.read("Dummy", KeySet.all(), Collections.singletonList("C"));
+              assertNotNull(resultSet);
+            } catch (IllegalStateException ex) {
+              // in case of multiplexed session we should allow concurrent requests on same session
+              // and should not receive a IllegalStateException
+              Assert.fail();
+            }
```

Comment:
nit: do we need this catch block? I think that the test would also fail without it if we get an uncaught IllegalStateException.

Reply:
I'm removing test and the current change in `AbstractReadContext` outside of this PR. I'll need to think a bit more on how `SessionTransaction` interface should be refactored to handle all use-cases.